### PR TITLE
Migrate git-repo manifest parsing from Jackson to KxS

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt [hunk 1]
```diff
@@ -38,7 +38,6 @@ import org.cyclonedx.model.Metadata
 import org.cyclonedx.model.license.Expression
 import org.cyclonedx.model.metadata.ToolInformation
 
-import org.ossreviewtoolkit.model.FileFormat
```
Comment (fviernau) [inline plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:41]: commit-msg: is ment -> is meant

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt [hunk 2]
```diff
@@ -356,17 +358,16 @@ class CycloneDxReporter : Reporter {
...
-            val outputFile = outputDir.resolve("$outputName.${fileFormat.fileExtension}")
+        validFileExtensions.forEach { fileExtension ->
+            val outputFile = outputDir.resolve("$outputName.$fileExtension")
 
-            val bomGenerator = when (fileFormat) {
-                FileFormat.XML -> BomGeneratorFactory.createXml(schemaVersion, bom)
-                FileFormat.JSON -> {
+            val bomGenerator = when (fileExtension) {
+                "xml" -> BomGeneratorFactory.createXml(schemaVersion, bom)
+                "json" -> {
```
Comment (fviernau) [inline plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:370]: IIUC this change is a functional change at least in the following aspects
1. If "unkown" file extension get specified, these get ignored. Before it failed.
2. The default is changed from xml to all formats.

Looking at this when case, the previous code did not redefine constants and had compiler check
whether all cases are covered by `when`.  What do you think about "reverting" that by simply defing a file extensions enum? (And keeping the code analog to before)

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt [hunk 3]
```diff
@@ -390,7 +391,7 @@ class CycloneDxReporter : Reporter {
                     BomGeneratorFactory.createJson(schemaVersion, bomWithoutExtensibleTypes) as Any
                 }
 
-                else -> throw IllegalArgumentException("Unsupported CycloneDX file format '$fileFormat'.")
+                else -> throw IllegalArgumentException("Unsupported CycloneDX file extension '$fileExtension'.")
```
Comment (fviernau) [inline plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:394]: What's the reason you choose to change the wording from "extension" to "format" ?

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt [hunk 4]
```diff
@@ -38,7 +38,6 @@ import org.cyclonedx.model.Metadata
 import org.cyclonedx.model.license.Expression
 import org.cyclonedx.model.metadata.ToolInformation
 
-import org.ossreviewtoolkit.model.FileFormat
 import org.ossreviewtoolkit.model.Identifier
 import org.ossreviewtoolkit.model.LicenseSource
 import org.ossreviewtoolkit.model.Package
```
Comment (fviernau) [inline plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:43]: commit-msg: Can you elaborate what  "This is a preparation for further cleanups." referrs to?

(I don't understand why this change is done)

File: gradle/libs.versions.toml [hunk 5]
```diff
@@ -61,6 +61,7 @@ springCore = "5.3.37"
 svnkit = "1.10.11"
 sw360Client = "17.0.1-m2"
 wiremock = "3.8.0"
+xmlutil = "0.90.1"
```
Comment (fviernau) [inline gradle/libs.versions.toml:64]: How does xmlutil compare to jackson for xml ? Is it similar powerful?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: gradle/libs.versions.toml (hunk 5 above)
Comment (sschuberth) [reply gradle/libs.versions.toml:64]: Can you be a bit more specific about what kind of features you have in mind?

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt (hunk 3 above)
Comment (sschuberth) [reply plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:394]: Because the code now really works directly with file extensions compared to semantic names for file formats (via the enum) compared to before.

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt (hunk 2 above)
Comment (sschuberth) [reply plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:370]: I was considering to introduce a separate file format enum for CycloneDX here, but then felt that 1) this case is so simple that it barely justifies the additional code, 2) that functionality should actually go upstream (I'm planning to make a contribution to the CycloneDX Java library so it "known" about its own formats that it supports).

Regarding the behavioral changes, you're correct. I'll change it back to fail again on an unknown format, and to default only to XML.

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt (hunk 4 above)
Comment (sschuberth) [reply plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:43]: This allows the last "Remove the unused XML mapper" commit to be made. Otherwise `FileFormat.XML` would depend on Jackson's XML mapper.

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt (hunk 2 above)
Comment (fviernau) [reply plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:370]: > so simple that it barely justifies the additional code

I was proposing the enum, because I thought it would actually safe code:
e.g. you could turn the input string into enum instance via clikt (as before) instead of adding logic which checks for valid values.

Comment (sschuberth) [reply plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:370]: > you could turn the input string into enum instance via clikt (as before)

Note that clikt was / is not involved here, as the reporters only see their reporter-specific options as part of the generic `PluginConfiguration`, and the `ReporterCommand` does not know about typed reporter-specific options.

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt [hunk 6]
```diff
@@ -148,10 +147,18 @@ class CycloneDxReporter : Reporter {
         val dataLicense = config.options[OPTION_DATA_LICENSE] ?: DEFAULT_DATA_LICENSE.id
         val createSingleBom = !config.options[OPTION_SINGLE_BOM].isFalse()
 
-        val outputFileFormats = config.options[OPTION_OUTPUT_FILE_FORMATS]
+        val outputFileExtensions = config.options[OPTION_OUTPUT_FILE_FORMATS]
             ?.split(",")
-            ?.mapTo(mutableSetOf()) { FileFormat.valueOf(it.uppercase()) }
-            ?: setOf(FileFormat.XML)
+            ?.mapTo(mutableSetOf()) { it.trim().lowercase() }
+            .orEmpty()
```
Comment (fviernau) [inline plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:153]: Instead of `orEmpty()` we can move the `ifEmpty` here. This removes the need for the  `validFileExtensions` variable entirely (remove lines 160 + 161) and to operate on ` outputFileExtensions` only.

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt [hunk 7]
```diff
@@ -356,17 +363,16 @@ class CycloneDxReporter : Reporter {
         schemaVersion: Version,
         outputDir: File,
         outputName: String,
-        requestedOutputFileFormats: Set<FileFormat>
+        validFileExtensions: List<String>
```
Comment (fviernau) [inline plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:366]: Propose to name this just `outputFileExtensions` for consistency with other parameters.

**Review by fviernau — CHANGES_REQUESTED** [review]

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt (hunk 2 above)
Comment (sschuberth) [reply plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:375]: > I'm planning to make a contribution to the CycloneDX Java library so it "known" about its own formats that it supports

FYI: https://github.com/CycloneDX/cyclonedx-core-java/pull/445

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt (hunk 6 above)
Comment (sschuberth) [reply plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt:153]: It was a tiny bit more trick as also the null case needs to be handled, but done.

**Review by fviernau — APPROVED** [review]

File: plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt (hunk 6 above)
Comment (fviernau) [reply plugins/reporters/cyclonedx/src/main/kotlin/CycloneDxReporter.kt]: Thanks!
